# Using KGTK query to do interesting queries in Wikidata
This notebook shows use cases of interesting queries on Wikidata that can be done using the KGTK query command (aka Kypher), and that cannot be done using the public Wikidata SPARQL endpoint

The notebook has a preamble to set up environment variables to access the relevant files

In [1]:
# Parameters

# Folder on local machine where to create the output and temporary folders
output_path = "/Users/pedroszekely/Downloads/kypher"

# The names of the output and temporary folders
output_folder = "wd-workshop"
temp_folder = "temp.wd-workshop"

# The location of input Wikidata files
wikidata_folder = "/Volumes/GoogleDrive/Shared drives/KGTK/datasets/wikidata-20210215/data/"
# wikidata_folder = "/Users/pedroszekely/Downloads/kypher/wikidata_os_v1/"
# The wikidata_os files can be downloaded from https://drive.google.com/drive/folders/1V6oAQKmwQ4LJnrBai-uv5gHWphFSCt50?usp=sharing

wikidata_dbpedia_folder = "/Volumes/GoogleDrive/Shared drives/KGTK/datasets/wikidata-dbpedia"

# Location of the cache database for kypher
cache_path = "/Users/pedroszekely/Downloads/kypher/temp.novartis"
# cache_path = "/Users/pedroszekely/Downloads/kypher/temp.useful_wikidata_files_v4/wikidata.sqlite3.db"
# Whether to delete the cache database
delete_database = False

# shortcuts to commands
kgtk = "time kgtk --debug"
# kgtk = "kgtk --debug"
# kgtk = "kgtk"

In [2]:
import io
import os
import subprocess
import sys
import time

import numpy as np
import pandas as pd

import altair as alt

import papermill as pm

In [3]:
start_time = round(time.time())

In [4]:
# The names of files in the KGTK Wikidata distirbution that we will use in this notebook.
file_names = {
    "claims": "claims.tsv.gz",
    "quantity": "claims.quantity.tsv.gz",
    "time": "claims.time.tsv.gz",
    "label": "labels.en.tsv.gz",
    "alias": "aliases.en.tsv.gz",
    "description": "descriptions.en.tsv.gz",
    "item": "claims.wikibase-item.tsv.gz",
    "external_id": "claims.external-id.tsv.gz",
    "qualifiers": "qualifiers.tsv.gz",
    "sitelinks": "sitelinks.tsv.gz",
    "qualifiers_time": "qualifiers.time.tsv.gz",
    "property_datatypes": "metadata.property.datatypes.tsv.gz",
    "isa": "derived.isa.tsv.gz",
    "p279star": "derived.P279star.tsv.gz",
    "p279": "derived.P279.tsv.gz",
    "p31": "derived.P31.tsv.gz",
    "dwd_isa": "derived.dwd_isa.tsv.gz"
}

# We will define environment variables to hold the full paths to the files as we will use them in the shell commands
kgtk_environment_variables = []

os.environ['WIKIDATA'] = wikidata_folder
kgtk_environment_variables.append('WIKIDATA')

for key, value in file_names.items():
    variable = key.upper()
    os.environ[variable] = wikidata_folder + value
    kgtk_environment_variables.append(variable)

os.environ["WD2DB"] = "/Volumes/GoogleDrive/Shared drives/KGTK/datasets/wikidata-dbpedia/wikidata_to_dbpedia_edge_file.tsv.gz"
kgtk_environment_variables.append("WD2DB")


# KGTK creates a SQLite database to index the knowledge graph.
if cache_path:
    os.environ['STORE'] = "{}/wikidata.sqlite3.db".format(cache_path)
else:
    os.environ['STORE'] = "{}/{}/wikidata.sqlite3.db".format(output_path, temp_folder)
kgtk_environment_variables.append('STORE')

# We will create many temporary files, so set up a folder for outputs and one for the temporary files.
os.environ['TEMP'] = "{}/{}".format(output_path, temp_folder) 
os.environ['OUT'] = "{}/{}".format(output_path, output_folder) 
kgtk_environment_variables.append('TEMP')
kgtk_environment_variables.append('OUT')

# Envronment variables with shortcuts to the commands we use often
os.environ['kgtk'] = kgtk
# Use for debugging, but careful as it causes import to dataframes to break
# os.environ['kypher'] = "time kgtk --debug query --graph-cache " + os.environ['STORE']
os.environ['kypher'] = "kgtk query --graph-cache " + os.environ['STORE']
kgtk_environment_variables.append('kgtk')
kgtk_environment_variables.append('kypher')

# We'll save the current working directory so we can call into other example notebooks later
os.environ["EXAMPLES_DIR"] = os.getcwd()
kgtk_environment_variables.append('EXAMPLES_DIR')

kgtk_environment_variables.sort()
for variable in kgtk_environment_variables:
    print("{}: \"{}\"".format(variable, os.environ[variable]))

ALIAS: "/Volumes/GoogleDrive/Shared drives/KGTK/datasets/wikidata-20210215/data/aliases.en.tsv.gz"
CLAIMS: "/Volumes/GoogleDrive/Shared drives/KGTK/datasets/wikidata-20210215/data/claims.tsv.gz"
DESCRIPTION: "/Volumes/GoogleDrive/Shared drives/KGTK/datasets/wikidata-20210215/data/descriptions.en.tsv.gz"
DWD_ISA: "/Volumes/GoogleDrive/Shared drives/KGTK/datasets/wikidata-20210215/data/derived.dwd_isa.tsv.gz"
EXAMPLES_DIR: "/Users/pedroszekely/Documents/GitHub/kgtk-at-2021-wikidata-workshop"
EXTERNAL_ID: "/Volumes/GoogleDrive/Shared drives/KGTK/datasets/wikidata-20210215/data/claims.external-id.tsv.gz"
ISA: "/Volumes/GoogleDrive/Shared drives/KGTK/datasets/wikidata-20210215/data/derived.isa.tsv.gz"
ITEM: "/Volumes/GoogleDrive/Shared drives/KGTK/datasets/wikidata-20210215/data/claims.wikibase-item.tsv.gz"
LABEL: "/Volumes/GoogleDrive/Shared drives/KGTK/datasets/wikidata-20210215/data/labels.en.tsv.gz"
OUT: "/Users/pedroszekely/Downloads/kypher/wd-workshop"
P279: "/Volumes/GoogleDrive/Shar

In [5]:
%cd {output_path}

/Users/pedroszekely/Downloads/kypher


Define the shortcuts for Kypher so that import the relevant files into the Kypher index and define shortcuts to make the queries nicer to write

In [6]:
!$kypher \
-i "$ITEM" --as items \
-i "$TIME" --as time \
-i "$P31" --as p31 \
-i "$P279" --as p279 \
-i "$LABEL" --as labels \
-i "$P279STAR" --as p279star \
-i "$EXTERNAL_ID" --as external_ids \
-i "$OUT"/ulan.tsv --as ulan \
--limit 10

id	node1	label	node2	rank	node2;wikidatatype
P10-P1629-Q34508-bcc39400-0	P10	P1629	Q34508	normal	wikibase-item
P10-P1855-Q15075950-7eff6d65-0	P10	P1855	Q15075950	normal	wikibase-item
P10-P1855-Q4504-a69d2c73-0	P10	P1855	Q4504	normal	wikibase-item
P10-P1855-Q69063653-c8cdb04c-0	P10	P1855	Q69063653	normal	wikibase-item
P10-P1855-Q7378-555592a4-0	P10	P1855	Q7378	normal	wikibase-item
P10-P2302-Q21502404-d012aef4-0	P10	P2302	Q21502404	normal	wikibase-item
P10-P2302-Q21510851-5224fe0b-0	P10	P2302	Q21510851	normal	wikibase-item
P10-P2302-Q21510852-dde2f0ce-0	P10	P2302	Q21510852	normal	wikibase-item
P10-P2302-Q52004125-d0288d06-0	P10	P2302	Q52004125	normal	wikibase-item
P10-P2302-Q53869507-974ce3b1-0	P10	P2302	Q53869507	normal	wikibase-item


## Retrieve large amounts of data from Wikidata

John is doing research on the popularity of first names to improve his entity resolution algorithm for people. He sees that Wikidata contains about 9 million people, so he wants to get the distribution of counts of first names from Wikidata. He writes a SPARQL query, but it times out, so he downloads the Wikidata KGTK files on his laptop and writes a kypher query. The query retrieves all instances of human (Q5), gets their frst names using the P735 property and return the counts.

John thinks he will want to do additional analysis on the data, so chooses standard KGTK names for the headers to generate the data as a KGTK graph that then he can use as input to other KGTK commands.

In [7]:
%%time
# compare to SPARQL
# paper: first names
!$kypher -i items -i p31 -i labels \
--match '\
    p31: (person)-[]->(:Q5), \
    items: (person)-[:P735]->(given_name), \
    labels: (given_name)-[]->(given_name_label)' \
--return 'distinct given_name as node1, count(given_name) as node2, given_name_label as `node1;label`, "count_names" as label' \
--order-by 'node2 desc' \
-o "$OUT"/given-names.tsv

CPU times: user 6.21 s, sys: 1.7 s, total: 7.91 s
Wall time: 8min 17s


In [8]:
!wc "$OUT"/given-names.tsv 

   53253  216696 1988754 /Users/pedroszekely/Downloads/kypher/wd-workshop/given-names.tsv


John takes a peek at the file to make sure he got the headers correcly: an edge from the q-node to the count, using `count_names` as the property, and including the `label` of `node1` so he can read the data. John sees that his name is by far the most popular name in Wikidata, and gets the information he needs to fine tune his entity resolution algorithms. 

In [9]:
!head "$OUT"/given-names.tsv | column -ts $'\t'

node1      node2   node1;label   label
Q4925477   120416  'John'@en     count_names
Q12344159  74235   'William'@en  count_names
Q4927937   59298   'Robert'@en   count_names
Q16428906  57107   'Thomas'@en   count_names
Q677191    52568   'James'@en    count_names
Q18057751  49005   'David'@en    count_names
Q2958359   44735   'Charles'@en  count_names
Q2793400   40987   'Peter'@en    count_names
Q1249148   40149   'Richard'@en  count_names


John gets curious and wants to know whether the popularity of names depends of time, so modifies his query to partition the data by people's year of birth.

In [10]:
!$kypher -i items -i time -i p31 -i labels \
--match '\
    p31: (person)-[]->(:Q5), \
    items: (person)-[:P735]->(given_name), \
    time: (person)-[:P569]->(date_of_birth), \
    labels: (given_name)-[]->(given_name_label)' \
--return 'distinct given_name as node1, kgtk_date_year(date_of_birth) as year, count(given_name) as node2, given_name_label as `node1;label`, "count_names_yearly" as label' \
--order-by 'given_name, cast(year, integer), node2 desc' \
-o "$OUT"/given-names.year.tsv

In [11]:
!wc "$OUT"/given-names.year.tsv 

  882179 4431755 42410662 /Users/pedroszekely/Downloads/kypher/wd-workshop/given-names.year.tsv


John takea a quick peek at the file to verify that the headers are correct.

In [12]:
!head -5 "$OUT"/given-names.year.tsv | column -ts $'\t'

node1     year  node2  node1;label      label
Q1000387  1798  1      'Ferdinanda'@en  count_names_yearly
Q1000387  1849  1      'Ferdinanda'@en  count_names_yearly
Q1000387  1868  1      'Ferdinanda'@en  count_names_yearly
Q1000387  1870  1      'Ferdinanda'@en  count_names_yearly


John heard anecdotaly that Jessica had become a popular name in the late 90s and greps for Jessica in the file. 

In [13]:
!grep "'Jessica'" "$OUT"/given-names.year.tsv | tail -15

Q630846	1995	58	'Jessica'@en	count_names_yearly
Q630846	1996	41	'Jessica'@en	count_names_yearly
Q630846	1997	27	'Jessica'@en	count_names_yearly
Q630846	1998	23	'Jessica'@en	count_names_yearly
Q630846	1999	23	'Jessica'@en	count_names_yearly
Q630846	2000	51	'Jessica'@en	count_names_yearly
Q630846	2001	18	'Jessica'@en	count_names_yearly
Q630846	2002	18	'Jessica'@en	count_names_yearly
Q630846	2003	11	'Jessica'@en	count_names_yearly
Q630846	2004	6	'Jessica'@en	count_names_yearly
Q630846	2005	2	'Jessica'@en	count_names_yearly
Q630846	2009	1	'Jessica'@en	count_names_yearly
Q630846	2011	1	'Jessica'@en	count_names_yearly
Q630846	2014	1	'Jessica'@en	count_names_yearly
Q630846	2016	2	'Jessica'@en	count_names_yearly


John realizes that he needs to normalize the counts of names by the number of people born in each year. He wonders whether he can do it in one kypher query, but takes the easy way out and writes a simple query to get the counts of people born each year. He can do this faster than he can think of a complex query to get the final result in one go.

In [14]:
!$kypher -i time -i p31 \
--match ' \
    p31: (person)-[]->(:Q5), \
    time: (person)-[:P569]->(date_of_birth)' \
--return 'kgtk_date_year(date_of_birth) as node1, count(person) as node2, "count_people_born" as label' \
-o "$TEMP"/human.count.year.tsv

John is happy that KGTK accepts literals as subjects of triples because here the subjects (`node1`) are years.

In [15]:
!head "$TEMP"/human.count.year.tsv

node1	node2	label
1	105	count_people_born
2	5	count_people_born
3	9	count_people_born
4	8	count_people_born
5	11	count_people_born
6	10	count_people_born
7	7	count_people_born
8	5	count_people_born
9	9	count_people_born


John knows he is almost there. He needs to get the names from the `given-names.year.tsv` file, and needs to pick out the year from the qualifier he put on the edge using the syntax to get the attributes of edges `[r {year: the_year}]`. He computes the fraction of people with each name and multiplies by 100,000 so that the numbers are not so tiny and easier to read. John also gets the labels of the q-nodes from the attribute he put on `node1` so tha the doesn't have to join with the `labels.tsv` file again.

In [16]:
!$kypher -i "$OUT"/given-names.year.tsv -i "$TEMP"/human.count.year.tsv \
--match ' \
    names: (given_name {label: given_name_label})-[r {year: the_year}]->(count_names), \
    year: (the_year)-[]->(count_people)' \
--return 'given_name as node1, "normalized_count_names_yearly" as label, cast(count_names, float) * 10000 / cast(count_people, float) as node2, the_year as year, given_name_label as `node1;label`' \
--order-by 'given_name, cast(the_year, integer), node2 desc' \
-o "$OUT"/given-names.year.normalized.tsv

In [17]:
!head "$OUT"/given-names.year.normalized.tsv | column -ts $'\t'

node1     label                          node2               year  node1;label
Q1000387  normalized_count_names_yearly  2.207505518763797   1798  'Ferdinanda'@en
Q1000387  normalized_count_names_yearly  1.1767474699929394  1849  'Ferdinanda'@en
Q1000387  normalized_count_names_yearly  0.6827336655970506  1868  'Ferdinanda'@en
Q1000387  normalized_count_names_yearly  0.6334726973267453  1870  'Ferdinanda'@en
Q1000387  normalized_count_names_yearly  0.536711034778875   1888  'Ferdinanda'@en
Q1000433  normalized_count_names_yearly  1.0892059688487092  1852  'Bud'@en
Q1000433  normalized_count_names_yearly  0.9004141905276427  1858  'Bud'@en
Q1000433  normalized_count_names_yearly  0.6082355087890031  1881  'Bud'@en
Q1000433  normalized_count_names_yearly  0.5845218611176058  1882  'Bud'@en


John greps the normalized file again. Jessica was not a popular name in the 60s and began to get popular in the late 70s. John satisfied his curiosity. The popularity of names is time dependent, but for now, John will work to integrate the aggregate data into his entity resolution algorithm.

In [18]:
!grep "'Jessica'" "$OUT"/given-names.year.normalized.tsv | tail -50

Q630846	normalized_count_names_yearly	1.5782828282828283	1960	'Jessica'@en
Q630846	normalized_count_names_yearly	1.6427682992654478	1961	'Jessica'@en
Q630846	normalized_count_names_yearly	0.6936416184971098	1962	'Jessica'@en
Q630846	normalized_count_names_yearly	2.2975301550832854	1963	'Jessica'@en
Q630846	normalized_count_names_yearly	2.7612232218872963	1964	'Jessica'@en
Q630846	normalized_count_names_yearly	3.534651365553644	1965	'Jessica'@en
Q630846	normalized_count_names_yearly	3.7685601587820012	1966	'Jessica'@en
Q630846	normalized_count_names_yearly	2.2576760987357014	1967	'Jessica'@en
Q630846	normalized_count_names_yearly	3.7844383893430216	1968	'Jessica'@en
Q630846	normalized_count_names_yearly	4.271034846619601	1969	'Jessica'@en
Q630846	normalized_count_names_yearly	5.083022704168078	1970	'Jessica'@en
Q630846	normalized_count_names_yearly	8.008208413623965	1971	'Jessica'@en
Q630846	normalized_count_names_yearly	5.099569086412198	1972	'Jessica'@en
Q630846	normalized_count_names

## Analytics on full Wikidata

Jessica is working with John on the entity resolution algorithm and her job is to use the number of instances of each class in Wikidata as a feature. The query that Jessica needs to write is simple as she just needs to count the number of instances of each class, summing up over the instances of all subclasses. She knows that there are over 1 million classes in Wikidata (entities with a P279 property), so she knows it will not run on the public SPARQL endpint. Jessica gets the SQLite database from John so that she does not have to wait the 2 or so hours to load it on her laptop, writes the query and goes for lunch as she knows it will take a while for it to run.

In [19]:
%%time
# compare to SPARQL
# paper: instances
!$kypher -i p31 -i p279star \
--match '\
    p31: (entity)-[]->(class), \
    p279star: (class)-[]->(super_class)' \
--return 'distinct super_class as node1, count(distinct entity) as node2, "entity_count" as label' \
--order-by 'node2 desc, node1' \
-o "$OUT"/class.count.tsv.gz

CPU times: user 1min 1s, sys: 17.6 s, total: 1min 19s
Wall time: 1h 22min 58s


After coming back from lunch, the file is ready, it contains data for 75K classes, she figures that the other classes don't have instances.

In [20]:
!zcat < "$OUT"/class.count.tsv.gz | wc

   75195  225585 1863081


In [21]:
!zcat < "$OUT"/class.count.tsv.gz | head -5 

node1	node2	label
Q35120	88859643	entity_count
Q99527517	74418826	entity_count
Q488383	73704542	entity_count
Q28813620	68227171	entity_count
zcat: error writing to output: Broken pipe


Jessica is curious about the data, so she writes a query to get the counts of different classes of film (Q11424). Jessica had been working with John, so she learned the trick to use the standard names for column headings so that she can use the output of previous queries as new graphs. She shudders to think that if she was using SPARQL she would have had to set up a new Wikidata SPARQL endpoint to be able to load her personal data in it, and to be extremely caeful to not make a mistake because deleting the data would have been a chore. Jessica had watched John make several mistakes when he was building the files for the names. John had simply fixed the queries and re-run the other queries that depended on the data he had just fixed.

In [56]:
%%time
# compare to SPARQL
# paper: film instances
!$kypher -i p279star -i labels -i "$OUT"/class.count.tsv.gz \
--match ' \
    p279star: (class)-[]->(:Q11424), \
    count: (class)-[]->(count), \
    labels: (class)-[]->(class_label)' \
--return 'class as node1, class_label as `node1;label`, count as node2' \
--order-by 'cast(count, integer) desc' \
--limit 10 \
| column -ts $'\t'

node1      node1;label               node2
Q11424     'film'@en                 314889
Q24862     'short film'@en           33733
Q506240    'television film'@en      17310
Q226730    'silent film'@en          17131
Q20667187  'silent short film'@en    16302
Q202866    'animated film'@en        9019
Q17517379  'animated short film'@en  4100
Q10590726  'video album'@en          1931
Q24869     'feature film'@en         1643
Q430525    'concert film'@en         1319
CPU times: user 18.1 ms, sys: 13.1 ms, total: 31.2 ms
Wall time: 1.49 s


Jessica now has the statistics she needs to work on her feature for the entity resolution algorithm. 

## Extract new graphs from Wikidata

Bill is working on a project to find networks of researchers working on specific topics. He wants to use publication data to find relationships among authors using publications. Bill knows that he can get lots of publication data from Pubmed or Microsoft Academic graph, but wants to give Wikidata a try as he heard that Wikidata has close to 40 million publications, and that in Wikidata publications have links to other entities such as main subjects.

Bill decides that the simplest experiment to try first is to build a network of authors of publications in Wikidata: he wants to create a graph of people in Wikidata who authored papers, to put a link between two people if the coauthored a paper, and to add a qualifier with the count of papers they coauthored. He knows the computation is expensive as there are 40ish million papers in Wikidata, so the network will be large. He doesn't even try to write a SPARQL query because he knows it will time out. Bill downloads the KGTK files and decides to write his first query using only 2019 data so he doesn't have to wait so long if he makes a mistake.

### First do it for 2019 to debug the query

In [23]:
!$kypher -i p31 -i p279star -i items -i time -i labels \
--match '\
    p31: (pub)-[]->(class), \
    p279star: (class)-[]->(:Q591041), \
    time: (pub)-[:P577]->(pub_date), \
    items: (pub)-[:P50]->(author1), \
    items: (pub)-[:P50]->(author2), \
    labels: (author1)-[]->(author1_label)' \
--where 'author1 < author2 and kgtk_date_year(pub_date) = 2019' \
--return 'distinct author1 as node1, "Pcoauthor" as label, author2 as node2, count(distinct pub) as count_publications, author1_label as `node1;label`' \
--order-by 'count_publications desc' \
-o "$TEMP"/coauthors.2019.id.tsv.gz

In [24]:
!zcat < "$TEMP"/coauthors.2019.id.tsv.gz | head

node1	label	node2	count_publications	node1;label
Q104625960	Pcoauthor	Q104626213	117	'Secundino López Puente'@en
Q104625960	Pcoauthor	Q42121517	117	'Secundino López Puente'@en
Q104625960	Pcoauthor	Q46702124	117	'Secundino López Puente'@en
Q104625960	Pcoauthor	Q57221019	117	'Secundino López Puente'@en
Q104625960	Pcoauthor	Q57235422	117	'Secundino López Puente'@en
Q104625960	Pcoauthor	Q62593499	117	'Secundino López Puente'@en
Q104625960	Pcoauthor	Q62607742	117	'Secundino López Puente'@en
Q104625960	Pcoauthor	Q80042771	117	'Secundino López Puente'@en
Q104626213	Pcoauthor	Q42121517	117	'Roberto Edoardo Villa'@en
zcat: error writing to output: Broken pipe


Bill wants to sanity check his data so he looks up the first person in Google Scholar and finds that Secundino López Puente has many publications in 2019. Looks like the query is working fine.

### Build the network for all authors

Bill removes the year restriction and runs the query for the full data. The query for a single year took close to 10 minutes, so Bill decides to leave the query running overnight.

In [25]:
%%time
# compare to SPARQL
!$kypher -i p31 -i p279star -i items \
--match '\
    p31: (pub)-[]->(class), \
    p279star: (class)-[]->(:Q591041), \
    items: (pub)-[:P50]->(author1), \
    items: (pub)-[:P50]->(author2)' \
--where 'author1 < author2' \
--return 'distinct author1 as node1, "Pcoauthor" as label, author2 as node2, count(distinct pub) as count_publications' \
--order-by 'count_publications desc' \
-o "$TEMP"/coauthors.tsv.gz

In [26]:
!zcat < "$TEMP"/coauthors.tsv.gz | head | column -ts $'\t'

zcat: error writing to output: Broken pipe
node1      label      node2      count_publications
Q67650927  Pcoauthor  Q84519428  1705
Q92189676  Pcoauthor  Q92232927  1666
Q92189676  Pcoauthor  Q92470745  1659
Q92232927  Pcoauthor  Q92602887  1653
Q92232927  Pcoauthor  Q92470745  1650
Q92189676  Pcoauthor  Q92602887  1647
Q92470745  Pcoauthor  Q92602887  1631
Q67732460  Pcoauthor  Q92602887  1618
Q67732460  Pcoauthor  Q92189676  1616


### Build a network of authors who authored papers about cancer

Bill is interested in cancer research, so he wants to build the same network but using only the papers about cancer. He knows Wikidata has an extensive class hiearchy, so he writes a query to peek at the hierarchy below the q-node for cancer.
He writes a query to retrieve subclasses of cancer.

In [27]:
!$kypher -i p279star -i labels \
--match '\
    p279star: (cancer_type)-[]->(:Q12078), \
    labels: (cancer_type)-[]->(cancer_type_label)' \
--return 'cancer_type as node1, cancer_type_label as node2' \
--limit 10

node1	node2
Q101541302	'pulmonary artery intimal sarcoma'@en
Q101541613	'rectal small cell carcinoma'@en
Q101541672	'CIC-DUX4 sarcoma'@en
Q101541689	'colorectal large cell neuroendocrine carcinoma'@en
Q1016605	'Burkitt lymphoma'@en
Q102258467	'diffuse gastric cancer'@en
Q102293219	'luminal breast carcinoma B'@en
Q102293292	'skin meningioma'@en
Q102293358	'breast implant-associated anaplastic large cell lymphoma'@en
Q102293373	'salivary gland mucinous adenocarcinoma'@en


The results are promising, so Bill now incorporates the query for types of cancer into the query for building the coauthor network. He just needs to get the main subject of the paper using the `P921` property and test that the main subject is a subclass of cancer. He expects the query to be much faster because now it has strong restriction, so he gives it a try.

In [61]:
%%time
# compare to SPARQL
!$kypher -i p31 -i p279star -i items -i labels \
--match '\
    p31: (pub)-[]->(class), \
    p279star: (class)-[]->(:Q591041), \
    items: (pub)-[:P50]->(author1), \
    items: (pub)-[:P50]->(author2), \
    items: (pub)-[:P921]->(cancer_type), \
    p279star: (cancer_type)-[]->(:Q12078), \
    labels: (author1)-[]->(author1_label), \
    labels: (author2)-[]->(author2_label)' \
--where 'author1 < author2' \
--return 'distinct author1 as node1, "Pcoauthor" as label, author2 as node2, count(distinct pub) as count_publications, author1_label as `node1;label`, author2_label as `node2;label`' \
--order-by 'count_publications desc' \
-o "$TEMP"/coauthors.cancer.tsv.gz

CPU times: user 1.83 s, sys: 509 ms, total: 2.34 s
Wall time: 2min 37s


The query takes less than a minute and produces a network with close to half a million edges. Bill takes a peek to see what is in it, and now wonders whether he could have written the query in SPARQL and run it on the public SPARQL endpoint.

In [29]:
!zcat < "$TEMP"/coauthors.cancer.tsv.gz | wc

  228007 1959466 17105798


In [30]:
!zcat < "$TEMP"/coauthors.cancer.tsv.gz | head | column -ts $'\t'

node_x     relation   node_y     count_publications  node1;label                node2;label
Q60320900  Pcoauthor  Q60394812  396                 'Jorge Eduardo Cortes'@en  'Hagop Kantarjian'@en
Q60394812  Pcoauthor  Q66370727  236                 'Hagop Kantarjian'@en      'Susan O\'Brien'@en
zcat: Q40614280  Pcoauthor  Q60394812  186                 'Farhad Ravandi'@en        'Hagop Kantarjian'@en
Q60394812  Pcoauthor  Q66385413  180                 'Hagop Kantarjian'@en      'Guillermo Garcia-Manero'@en
Q60320900  Pcoauthor  Q66370727  172                 'Jorge Eduardo Cortes'@en  'Susan O\'Brien'@en
Q28958315  Pcoauthor  Q42748966  150                 'Hermann Brenner'@en       'Michael Hoffmeister'@en
Q60394812  Pcoauthor  Q64026412  150                 'Hagop Kantarjian'@en      'Stefan Faderl'@en
Q60394812  Pcoauthor  Q66370888  150                 'Hagop Kantarjian'@en      'Elias Jabbour'@en
error writing to outputQ60320900  Pcoauthor  Q66385413  146                 'Jorge Edu

Bill puts the first two names in Google and finds that they are famous and have publshied a lot together. Bill is happy to have a network with close to half a million edges that he can use to do interesting analyses.

## Combining Wikidata with external files

Abigail is working on a cultural heritage project, collaborating with the Getty Research Institute who gave her a file with 27 thousand ULAN identifiers. Abigail has a database indexed using VIAF identifiers, and wants to map her ULAN identifiers to VIAF identifiers so that she can use her database. She puts one of the ULAN identifiers in the Wikidata search box and discovers that Wikidata has both ULAN and VIAF identifiers for many artists. Abigail knows a little bit of SPARQL and easity figures out that it is easy to write a query to retrieve the VIAF identifier given a ULAN identifier. Her solution would require sending 27,000 queries to Wikidata, which would involve writing a Python script.

In [31]:
!wc "$OUT"/ulan.tsv

   27415   27415  356389 /Users/pedroszekely/Downloads/kypher/wd-workshop/ulan.tsv


Her colleague Bill tells her that she can easily solve the problem using KGTK query. The only thing she needs to do is to rename the header of her file with identifiers to `node1` and write a Kypher query.

In [58]:
%%time
# paper: ULAN ids
# compare to SPARQL
!$kypher -i items -i external_ids -i labels -i ulan \
--match '\
    ulan: (ulan_id)-[]->(), \
    external_ids: (viaf_id)<-[:P214]-(artist)-[:P245]->(ulan_id), \
    labels: (artist)-[]->(artist_label)' \
--return 'artist as node1, viaf_id as `node1;P214`, ulan_id as `node1;P245`, artist_label as `node1;label`' \
-o "$OUT"/ulan-to-viaf.tsv

CPU times: user 149 ms, sys: 49 ms, total: 198 ms
Wall time: 11.8 s


Abigail is thrilled to see that the query ran in less than 30 seconds and is curious to see the results. She got matches for 8,116 ULAN ids, which means that now she can get a lot of data from her database to do her analysis.

In [59]:
!wc "$OUT"/ulan-to-viaf.tsv

    8116   42730  443909 /Users/pedroszekely/Downloads/kypher/wd-workshop/ulan-to-viaf.tsv


In [60]:
!head "$OUT"/ulan-to-viaf.tsv | column -ts $'\t'

node1       node1;P214   node1;P245   node1;label
Q1000596    "20822441"   "500072302"  'Noémi Ferenczy'@en
Q1001063    "96418002"   "500099612"  'Olga Fialka'@en
Q100156272  "309815799"  "500335625"  'Gloria López Córdova'@en
Q100249806  "184467129"  "500040990"  'Alice Denniston Laughlin'@en
Q100250000  "63899160"   "500034511"  'Shirley L. Bolton'@en
Q100278786  "309815915"  "500336052"  'Winifred Casson'@en
Q100323915  "95510425"   "500332031"  'Claudia Müller'@en
Q100348403  "95887586"   "500033567"  'Priscilla Kepner Sage'@en
Q100377312  "233761"     "500288751"  'Cristina Castel-Branco'@en


## Combining Wikidata with DBpedia

After mining her VIAF database, Abigail realizes that there is a lot of interesting data in Wikipedia infoboxes that she would like to use in her analysis. She hears from a colleague that DBpedia extracts data from Wikipedia infoboxes. She is curious whether Wikidata already has most of this data. She browses the pages for some artists in Wikipedia and sees that the Wikipedia infoboxes have interesting information that she may want to include in her dataset.

Abigail downloads the DBpedia infobox data in RDF format and uses KGTK to convert it into KGTK format and to substitute the DBpedia URIs with Wikidata Q-nodes. 

In [35]:
!zcat < "$OUT"/wikidata_infobox.tsv.gz | wc -l

 99251608


Abigail sees that there are almost 100 million edges in the DBpedia infobox graph, so she first adds the dataset to the kypher index.
> This operation is similar to loading the triples into a triple store to enable running queries. In kypher it is not necessary to explicitly load the file as kypher will automatically load the file and build indices the first time the file is used. Abigail is doing separately as she is curious to see how long it takes to load the file in kypher.

In [36]:
!$kypher -i "$OUT"/wikidata_infobox.tsv.gz --as infobox \
--limit 1

node1	label	node2	id
nodemxZbyK2VRrGoaxfdLmyLxw-1	dbpedia:structured_value	"2019-08-07"	nodemxZbyK2VRrGoaxfdLmyLxw-1-dbpedia:structured_value-944f0e


Abigail first wants to see which properties are available for the people she has in the ULAN file. She constructs a query to count the number of statements for each property for the artists in her ULAN file.

> This query combines Wikidata with two external sources, her ULAN identifiers and DBPedia infoboxes.

In [37]:
!$kypher -i infobox -i ulan -i external_ids \
--match '\
    ulan: (ulan_id)-[]->(), \
    external_ids: (artist)-[:P245]->(ulan_id), \
    infobox: (artist)-[l]->()' \
--return 'l.label as node1, count(distinct l) as node2' \
--order-by 'node2 desc' \
--limit 20 \
| column -ts $'\t'

node1                  node2
property:birthPlace    3725
property:name          3208
property:birthDate     2738
property:nationality   2464
property:spouse        2265
property:deathPlace    1967
property:deathDate     1955
property:field         1645
property:caption       1494
property:birthName     1139
property:knownFor      1065
property:education     972
property:training      924
property:movement      803
property:occupation    785
property:awards        687
property:almaMater     557
property:works         308
property:children      278
property:notableWorks  273


Abigail is interested in the information about spouses as she is thinking of doing an analysis on the occupation of spouses. She constructs a query to retrieve the spouse statements already present in Wikidata for her ULAN artists.

In [38]:
!$kypher -i items -i ulan -i external_ids -i labels \
--match '\
    ulan: (ulan_id)-[]->(), \
    external_ids: (artist)-[:P245]->(ulan_id), \
    items: (artist)-[l:P26]->(spouse)' \
--return 'artist as node1, l.label as label, spouse as node2' \
-o "$OUT"/spouses.ulan.wikidata.tsv

In [39]:
!wc "$OUT"/spouses.ulan.wikidata.tsv

    1342    4026   29557 /Users/pedroszekely/Downloads/kypher/wd-workshop/spouses.ulan.wikidata.tsv


She also constructs a query to count the spouse statements of ULAN artists in the DBpedia dataset an converts the DBpedia property to `P26`, the Wikidata property.

In [40]:
!$kypher -i items -i infobox -i ulan -i external_ids -i labels \
--match '\
    ulan: (ulan_id)-[]->(), \
    external_ids: (artist)-[:P245]->(ulan_id), \
    infobox: (artist)-[:`property:spouse`]->(spouse)' \
--return 'artist as node1, "P26" as label, spouse as node2' \
-o "$OUT"/spouses.ulan.dbpedia.tsv

Abigail is encouraged as there are almost 1,000 additional statements in DBpedia that she may be able to import into her dataset.

In [41]:
!wc "$OUT"/spouses.ulan.dbpedia.tsv

    2270    8022   57656 /Users/pedroszekely/Downloads/kypher/wd-workshop/spouses.ulan.dbpedia.tsv


Abigail worries that some of the spouses may be strings rather than entities as she has seen Wikipedia infoboxes where some values are links and others are strings. She uses the regex feature in Kypher to count the number of spouses that are Wikidata q-nodes. 

In [42]:
!$kypher -i "$OUT"/spouses.ulan.dbpedia.tsv \
--match '()-[]->(spouse)' \
--where 'spouse =~ "^Q[0-9]+"' \
--return 'count(distinct spouse) as count_spouses_with_qnodes'

count_spouses_with_qnodes
449


Abigail is disappointed to see that only 449 are q-nodes, so uses grep to see what else is in the file. She sees that the DBpedia data is noisy as there are empty strings, numbers that look like dates and entities that do not correspond to Wikidata entities. 

In [43]:
!grep -v '\tQ[0-9]' "$OUT"/spouses.ulan.dbpedia.tsv | head

node1	label	node2
Q235186	P26	''@en
Q235186	P26	nodemxZbyK2VRrGoaxfdLmyLxw-3223237
Q235186	P26	nodemxZbyK2VRrGoaxfdLmyLxw-3223238
Q466241	P26	''@en
Q466241	P26	1985
Q466241	P26	'Patrick Robyn'@en
Q534385	P26	''@en
Q534385	P26	2010
Q3816460	P26	''@en


Abigail does not want to take on what looks like a difficult data cleaning and entity linking job, so she keeps the 449 clean entities and puts them in `spouses.ulan.dbpedia.qnodes.tsv`.

In [44]:
!grep '\tQ[0-9]' "$OUT"/spouses.ulan.dbpedia.tsv > "$OUT"/spouses.ulan.dbpedia.qnodes.tsv

Abigail wants to see how many artists have spouse statements in both Wikidata and DBpedia, so she adds another clause to the query and sees that only 359 are in both, so she can get about 100 new statements from DBpedia. While not a lot, she will diff the two files and keep the new statements (not shown in this notebook).

In [45]:
!$kypher -i items -i infobox -i ulan -i external_ids -i labels \
--match '\
    ulan: (ulan_id)-[]->(), \
    external_ids: (artist)-[:P245]->(ulan_id), \
    infobox: (artist)-[l:`property:spouse`]->(spouse), \
    items: (artist)-[:P26]->(spouse), \
    labels: (spouse)-[]->(spouse_label)' \
--return 'count(distinct l)' 

count(DISTINCT graph_12_c3."id")
359


### spouse statements in DBpedia
Abigail gets curious wether it is worth working on a project to augment Wikidata with spouse statements from DBpedia. Now she knows that she needs to focus on the ones that are mapped to q-nodes, so writes a query to fetch all the spouse statements from DBpedia. 

In [46]:
# paper: dbpedia spouses
!$kypher -i infobox -i p31 \
--match ' \
    infobox: (artist)-[:`property:spouse`]->(spouse)' \
--where 'spouse =~ "^Q[0-9]+"' \
--return 'artist as node1, "P26" as label, spouse as node2' \
| wc -l

   41211
CPU times: user 49.8 ms, sys: 23.1 ms, total: 72.9 ms
Wall time: 3.62 s


Abigail wants to make sure to get clean data, so she adds a constraint to verify that the q-nodes she gets from DBpedia are instances of `Q5`  (human) in Wikidata.

In [63]:
%%time
# paper: dbpedia spouses
!$kypher -i infobox -i p31 -i labels \
--match ' \
    infobox: (artist)-[:`property:spouse`]->(spouse), \
    p31: (spouse)-[]->(:Q5)' \
--opt 'labels: (spouse)-[:label]->(spouse_label)' \
--return 'artist as node1, "P26" as label, spouse as node2, spouse_label as `node2;label`' \
-o "$OUT"/spouses.dbpedia.qnodes.tsv

CPU times: user 2.56 s, sys: 714 ms, total: 3.27 s
Wall time: 3min 26s


Abigail sees that testing that the spouses are instances of `Q5` reduces the number of statements, so some of the URIs from DBpedia are either incorrect or incorrectly mapped to Wikidata. She wants to play it safe, so keeps the data that verifies that the spouses are human.

In [48]:
!wc -l "$OUT"/spouses.dbpedia.qnodes.tsv

   39778 /Users/pedroszekely/Downloads/kypher/wd-workshop/spouses.dbpedia.qnodes.tsv


In [65]:
!head "$OUT"/spouses.dbpedia.qnodes.tsv | column -ts $'\t'

node1     label  node2     node2;label
Q268177   P26    Q1000505  'Bud Lee'@en
Q673856   P26    Q1000682  'Fernando Carrillo'@en
Q1325720  P26    Q1000874  'Thomas Montacute, 4th Earl of Salisbury'@en
Q264908   P26    Q1001     'Mahatma Gandhi'@en
Q8250426  P26    Q10011    'Jiajing Emperor of Ming'@en
Q3784373  P26    Q10011    'Jiajing Emperor of Ming'@en
Q4834024  P26    Q1001114  'Buddy Baker'@en
Q1707377  P26    Q1001130  'Buddy Bregman'@en
Q33941    P26    Q1001933  'Helena'@en


Abigail also gets all the spouse statements from Wikidata

In [49]:
!$kypher -i items \
--match '(artist)-[l:P26]->(spouse)' \
--return 'artist as node1, l.label as label, spouse as node2' \
-o "$OUT"/spouses.wikidata.tsv

In [50]:
!wc -l "$OUT"/spouses.wikidata.tsv 

  681174 /Users/pedroszekely/Downloads/kypher/wd-workshop/spouses.wikidata.tsv


Abigail sees that there is a significant difference, so writes a query to identify the common statements in both datasets and sees that there are about clean 7,000 spouse statements in DBpedia that are not present in Wikidata. It will be easy to add them using Wikidata quick statements, a project that she will try to do later.
> Kypher can run queries over the two new graphs created in the previous queries

In [51]:
!$kypher -i "$OUT"/spouses.wikidata.tsv  -i "$OUT"/spouses.dbpedia.qnodes.tsv \
--match '\
    wikidata: (person)-[]->(spouse), \
    dbpedia: (person)-[]->(spouse)' \
--return 'distinct person as person, spouse as spouse' \
| wc -l

   32453


## Validate property constraints in Wikidata

Amir and Sarah are starting a project to find constraint violations in Wikidata. They find that constraints are associated with properties using the `P2302` (property constraint) property. There are over 44,000 constraints, so finding violations is a dauting task as many constraints apply to a very large number of statements.

In [52]:
!$kypher -i items \
--match '(property)-[l:P2302]->(constraint)' \
--return 'count(distinct l) as count' 

count
44552


Sarah refines the query to print the counts of the different types of constraints, and they see that 9 constraint types are significantly more popular than the others. Amir and Sarah decide to focus on the `value type constraint` as this is the constraint that checks that the value of a statement belongs to specific classes. This constraint is defined for 964 properties, so it is worth working on it.

In [53]:
!$kypher -i items -i labels -i p279star \
--match ' \
    items: (property)-[l:P2302]->(constraint), \
    labels: (constraint)-[]->(constraint_label)' \
--return 'constraint_label as constraint_label, count(distinct l) as count' \
--order-by 'count desc' \
| column -ts $'\t'

constraint_label                                        count
'item requires statement constraint'@en                 7576
'allowed entity types constraint'@en                    6401
'format constraint'@en                                  6042
'distinct values constraint'@en                         5513
'single value constraint'@en                            5453
'type constraint'@en                                    5070
'property scope constraint'@en                          2908
'conflicts-with constraint'@en                          1275
'value type constraint'@en                              964
'allowed qualifiers constraint'@en                      573
'allowed units constraint'@en                           483
'required qualifier constraint'@en                      391
'range constraint'@en                                   327
'value requires statement constraint'@en                320
'citation needed constraint'@en                         284
'one-of constraint'@en        

In [54]:
!$kypher --show-cache

Graph Cache:
DB file: /Users/pedroszekely/Downloads/kypher/temp.novartis/wikidata.sqlite3.db
  size:  142.07 GB   	free:  0 Bytes   	modified:  2021-07-25 21:24:24

KGTK File Information:
/Users/pedroszekely/Downloads/kypher/temp.wd-workshop/human.count.year.tsv:
  size:  50.79 KB   	modified:  2021-07-25 19:00:52   	graph:  graph_17
/Users/pedroszekely/Downloads/kypher/wd-workshop/class.count.tsv.gz:
  size:  322.36 KB   	modified:  2021-07-25 20:24:03   	graph:  graph_18
/Users/pedroszekely/Downloads/kypher/wd-workshop/given-names.year.tsv:
  size:  40.45 MB   	modified:  2021-07-25 18:57:56   	graph:  graph_16
/Users/pedroszekely/Downloads/kypher/wd-workshop/spouses.dbpedia.qnodes.tsv:
  size:  826.70 KB   	modified:  2021-07-25 21:23:33   	graph:  graph_15
/Users/pedroszekely/Downloads/kypher/wd-workshop/spouses.ulan.dbpedia.tsv:
  size:  56.30 KB   	modified:  2021-07-25 21:21:12   	graph:  graph_19
/Users/pedroszekely/Downloads/kypher/wd-workshop/spouses.wikidata.tsv:
  size:  15

In [55]:
"Took {:.2f} minutes to run the notebook from start to end".format((round(time.time()) - start_time)/60)

'Took 164.55 minutes to run the notebook from start to end'